### PR TITLE
Interface definition for application defined thread implementation.

### PR DESCRIPTION
**Pull request by temeo**
The interface will allow application to inject service interface into
provider. The service interface will define callbacks to manage
threads, mutexes and conditions variables.

### CODE REVIEW

File: wsrep_thread_service.h [hunk 1]
```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright (C) 2019 Codership Oy <info@codership.com>
+ *
+ * This file is part of wsrep-API.
+ *
+ * Wsrep-API is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
```
Comment (ayurchen) [inline wsrep_thread_service.h]: If we want to have a separate wsrep_loader.h file, it makes sense to put the same wording in there. And probably in wsrep_api.h as well. I mean, to have a common head in all three files.

File: wsrep_thread_service.h [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+    /** Thread type tags */
+    typedef struct wsrep_thread_key_st wsrep_thread_key_t;
+    typedef struct wsrep_thread_st wsrep_thread_t;
+    typedef struct wsrep_mutex_key_st wsrep_mutex_key_t;
+    /** Mutex type tags */
+    typedef struct wsrep_mutex_st wsrep_mutex_t;
+    typedef struct wsrep_cond_key_st wsrep_cond_key_t;
+    /** Condition variable tags */
+    typedef struct wsrep_cond_st wsrep_cond_t;
```
Comment (ayurchen) [inline wsrep_thread_service.h]: I think here line 63 and line 64 must be swapped. And same goes for 66 and 67

**Comment by temeo** [comment]
> What is the motivation to have a separate `wsrep_loader.h`?

This change was not necessary at this point. Reverted the changes to `wsrep_api.h` and removed `wsrep_loader.h`.

File: wsrep_thread_service.h (hunk 1 above)
Comment (temeo) [reply wsrep_thread_service.h]: Maybe better to update the license header for all files in separate PR.

**Pull request merged**